mm2src/mm2_main/src/lp_swap/taker_swap.rs

```diff
@@ -1905,86 +1906,80 @@ impl TakerSwap {
+                        // Roll back to confirming the maker payment spend.
+                        RecoveredSwapAction::SpentOtherPayment => {
+                            info!("Refund canceled. Maker payment spend tx {:02x}", tx_ident.tx_hash);
+                            // TODO: We prepared for refund but didn't finalize refund. This must be breaking something for lightning.
+                            // We better find a way to rollback the state machine and remove erroneous events,
+                            // the swap at this point will be marked as errored but in fact it recovered from the error.
+                            Ok((Some(TakerSwapCommand::ConfirmMakerPaymentSpend), vec![
+                                TakerSwapEvent::MakerPaymentSpent(tx_ident),
+                            ]))
+                        },
+                    };
+                },
+                // Encountered an error during swap recover.
+                Err(err) => match err.into_inner() {
+                    // The payment tx we want to refund isn't even on-chain. There is nothing to refund/spend.
+                    RecoverSwapError::PaymentTxNotFound => {
+                        return Ok((Some(TakerSwapCommand::Finish), vec![
+                            TakerSwapEvent::TakerPaymentRefundFailed(
+                                "TakerPayment isn't even on-chain to refund it.".into(),
+                            ),
+                        ]));
+                    },
+                    // The error is unrecoverable, retrying will not fix the issue.
+                    RecoverSwapError::Irrecoverable(e) => {
+                        return Ok((Some(TakerSwapCommand::Finish), vec![
+                            TakerSwapEvent::TakerPaymentRefundFailed(ERRL!("!taker_coin.recover_funds: {}", e).into()),
+                        ]));
+                    },
+                    // The error is temporary, retrying may fix the issue.
+                    RecoverSwapError::Temporary(e) => {
+                        error!("Error {} on recover_funds, retrying in 30 seconds", e);
+                        Timer::sleep(30.).await;
+                    },
+                    // We should wait for this many seconds and try again.
+                    RecoverSwapError::WaitAndRetry(secs) => {
+                        Timer::sleep(secs as f64).await;
+                    },
+                    // The swap will be automatically recovered after the specified locktime.
+                    RecoverSwapError::AutoRecoverableAfter(locktime) => {
+                        let taker_payment = self.r().taker_payment.as_ref().unwrap().tx_hex.clone();
+                        // FIXME: We should try to `recover_funds` again after locktime. The taker payment might have been
+                        //        spent by the maker at this point and we should go for spending the maker payment instead.
```

Comment:
Well, `wait_for_htlc_refund` for lightning doesn't depend on locktime, but this is a general code and we should handle this case. For the lightning case, if we got payment successful from here https://github.com/KomodoPlatform/komodo-defi-framework/blob/f401497aee55ee8152f96955bd40dd147b8ab88f/mm2src/coins/lightning.rs#L824-L828
It means that we should be spending the maker payment.

Reply:
i think that makes sense in the general sense. if the `wait_for_htlc_refund` fails we can try to recover again.

Reply:
there is one case that comes to my mind where it would be bad to retry endlessly: if the initial tx was reverted because of too low gas (so basically a misconfiguration from our side)... is you retry that tx, you will spend gas again, but the result will be the same if you haven't increased gas limits... so you loose the txfees every time you try
idk if we want/should account for this case, but thought i mention it

Reply:
yeah that's a good reason to have a `start/stop` rpc.

Reply:
A start/stop rpc doesn't solve this, as too much gas might have been already spent until the user or @cipig is aware of this, this could drain the user's wallet which is really bad. We need to mark swap as failed for this and this is one of the situations where we will still require manual recover funds usage. @cipig are you aware of other situations like this one, we shouldn't release this PR/fix unless we are sure it's completely safe to retry forever and all edge cases are covered.

Reply:
there is this https://github.com/KomodoPlatform/komodo-defi-framework/issues/1567, but there you don't loose any fees when you retry forever
i guess reverted "out of gas" txes on EVM chains are the only problem where you loose money if you retry forever
there may be other reasons why a EVM chain reverts your tx, but i haven't encountered any other that would always fail
so short answer is "no", just this case :-)

Reply:
covered here: [a28c768](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2280/commits/a28c768abbd9c3228428f3c37ba3798214619eba)

for the gas issue with reverted txs: what about retrying the refund in an exponential backoff manner? this greatly reduces the amount of failed recoveries and doesn't hurt the perf/speed that much.

Reply:
Strictly speaking, 'reverted' means that tx was cancelled due to errors during the contract execution (for e.g. locktime has not been passed yet), so the tx may be retried yet. 
A tx can be cancelled due to out of gas. To prevent retries in this case, I think, we may analyse the tx receipt in eth code and return some unrecoverable error to the swap code.